controllers/installation_controller.go

```diff
@@ -249,6 +256,181 @@ func (r *InstallationReconciler) ReconcileK0sVersion(ctx context.Context, in *v1
+
+}
+
+// ReconcileHelmCharts reconciles the helm charts from the Installation metadata with the clusterconfig object.
+func (r *InstallationReconciler) ReconcileHelmCharts(ctx context.Context, in *v1beta1.Installation) error {
+  log := ctrl.LoggerFrom(ctx)
+	var clusterconfig k0sv1beta1.ClusterConfig
+
+  // skip if there's no config in the installer spec
+	if in.Spec.Config == nil {
+    log.Info("addons","configcheck","no config")
+    if in.Status.State == v1beta1.InstallationStateKubernetesInstalled {
+      in.Status.SetState(v1beta1.InstallationStateInstalled, "Installed")
+    }
+		return nil
+	}
+
+	meta, err := release.MetadataFor(ctx, in.Spec.Config.Version)
+	if err != nil {
+		return fmt.Errorf("failed to get release bundle: %w", err)
+	}
+
+	// skip if the new release has no addon configs
+	if meta.Configs == nil {
+    log.Info("addons","configcheck","no addons")
+    if in.Status.State == v1beta1.InstallationStateKubernetesInstalled {
+      in.Status.SetState(v1beta1.InstallationStateInstalled, "Installed")
+    }
+		return nil
+	}
+
+	if in.Status.State == v1beta1.InstallationStateKubernetesInstalled || in.Status.State == v1beta1.InstallationStateHelmChartUpdateFailure {
+    var installedCharts k0shelm.ChartList
+    if err := r.List(ctx, &installedCharts); err != nil {
+      return fmt.Errorf("failed to list nodes: %w", err)
+    }
+
+    targetCharts := meta.Configs.Charts
+
+    chartErrors := []string{}
+    chartDrift := false
+    // grab the installed charts
+    for _,chart := range installedCharts.Items {
+      // extract any errors from installed charts
+      if chart.Status.Error != "" {
+        chartErrors = append(chartErrors, chart.Status.Error)
+      }
+      // check for version drift between installed charts and charts in the installer metadata
+      for _,targetChart := range targetCharts {
+        if targetChart.Name != chart.Status.ReleaseName {
+          continue
+        }
+        if targetChart.Version != chart.Spec.Version {
+          chartDrift = true
+        }
+      }
+    }
+
+    // If any chart has errors, update installer state and return
+    if len(chartErrors) > 0 {
+      in.Status.SetState(v1beta1.InstallationStateHelmChartUpdateFailure,strings.Join(chartErrors,","))
+      return nil
+    }
+
+    // If all addons match their target version, mark installation as complete
+    log.Info("addons","chartdrift",chartDrift)
+    if chartDrift { 
+      in.Status.SetState(v1beta1.InstallationStateInstalled,"Addons upgraded")
+      return nil
+    }
+
+  }
+
+	// skip if installer is already complete
+	if in.Status.State == v1beta1.InstallationStateInstalled {
+		return nil
+	}
+
+	// We want to skip and requeue if the k0s upgrade is still in progress
+	if in.Status.State != v1beta1.InstallationStateKubernetesInstalled {
+		return nil
+	}
+
+	// fetch the current clusterconfig
+	if err := r.Get(ctx, client.ObjectKey{Name: "k0s", Namespace: "kube-system"}, &clusterconfig); err != nil {
```

Comment:
the whitespace from here on looks really odd
you might need to run fmt

Reply:
go fmt doesn't affect these lines. I've tidied it up a little though